store/tikv/region_cache.go

```diff
@@ -709,8 +737,110 @@ func (r *Region) ContainsByEnd(key []byte) bool {
 		(bytes.Compare(key, r.meta.GetEndKey()) <= 0 || len(r.meta.GetEndKey()) == 0)
 }
 
-// Store contains a tikv server's address.
+// storeLoader loads the Store info given by storeId.
+type storeLoader func(ctx context.Context, id uint64) (*metapb.Store, error)
+
+// Store contains a kv server's address.
 type Store struct {
-	ID   uint64
-	Addr string
+	sync.Mutex              // mutex to avoid duplicate load requests
+	StoreLoader storeLoader // loader to get store address from PD
+	lastLoadTS  int64       // last load success timestamp(sec)
+	addr        string      // loaded store address
+
+	ID uint64 // store id
+
+	accessibility unsafe.Pointer // store's accessibility, point to *storeAccess
+}
+
+// storeAccess contains store's access info.
+type storeAccess struct {
```

Comment:
We can use a single uint64 to store both failAttempt(uint32) and lastFailedTime(uint32).
Then we don't need to allocate memory to update it.

Reply:
We can also add an address changed flag in this field.
Then we don't need to resolve the address of a store, the address in a store is immutable, we can safely access it concurrently.
If we detect that the store address has been changed, we can reload the store from the store map.
And when we got a new store address from PD, we mark the old store access flag to address changed, then delete the old store from the store map.

Reply:
> We can also add an address changed flag in this field.
> Then we don't need to resolve the address of a store, the address in a store is immutable, we can safely access it concurrently.
> If we detect that the store address has been changed, we can reload the store from the store map.
> And when we got a new store address from PD, we mark the old store access flag to address changed, then delete the old store from the store map.

this be handle by 

```
	lastLoadTS  int64       // last load success timestamp(sec)
	addr        string      // loaded store address
```

addr only resolve at the first time, and we cannot detect store address change from some watcher? 
so just period  re-resolve addr via `lastLoadTS` 
at last, use atomic variable cannot handle load-failure, you can not `cas-load` but `read-load-cas if sucess`, then it will make many duplicate resolve request in high concurrent situation, so use mutex in store level is better.